### ISSUE TITLE
Shared modules in Python 2.7 cause errors related to restricted mode

### ISSUE THREAD

**Issue by bsteffensmeier** [ninia/jep#68]
When using shared modules, certain functionality is not available because the shared modules are being used in [restricted mode](https://docs.python.org/2/library/restricted.html). For example this java code will generate an exception.

```java
try(Jep jep = new Jep(new JepConfig().addSharedModules("xml.etree.ElementTree"))){
    jep.eval("import xml.etree.ElementTree as ET");
    jep.eval("ET.parse('fakeFile.xml')");
}
```

```
Exception in thread "main" jep.JepException: <type 'exceptions.IOError'>: file() constructor not accessible in restricted mode
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:647)
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:1182)
```

This issue was originally reported under #62 and also reported in the comments for #55. The two limitations that we have come across so far is that you cannot open files in restricted mode and you cannot unmarshal code objects. I have looked through the cpython source and most of the other checks for restricted mode are complex reflection that I would not expect to come up. The only other code I saw that might cause problems is related to pickling.

The shared modules are not deliberately restricted, it is only a side affect of how they were implemented. The shared modules keep the \_\_builtins\_\_ from the "top interpreter" so when they are used in other jep interpreters the \_\_builtins\_\_ do not match the interpreter and python assumes restricted mode.

The current best solution for this problem is to use python 3, which eliminated restricted mode. Another alternative is to avoid using shared modules which comes with extra limitations on closing jep interpreters.

**Comment by bsteffensmeier** [comment]
I've been investigating possible solutions. It is really easy to fix in a single Jep instance, for example if xml.etree.ElementTree is a shared module then I can do this:

```python
>>> import xml.etree.ElementTree as ET
>>> ET.parse('fakeFile.xml')
jep.JepException: <type 'exceptions.IOError'>: file() constructor not accessible in restricted mode
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:647)
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:1182)
>>> ET.__builtins__ = __builtins__
>>> ET.parse('fakeFile.xml')
jep.JepException: <type 'exceptions.IOError'>: (2, 'No such file or directory', 'fakeFile.xml')
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:647)
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:1182)
```

Im counting it as success that it manages to throw an IOError since that indicates it made it into the open() call. Another solution is to specifically target the open function instead of replacing all the __builtins__

```
>>> import xml.etree.ElementTree as ET
>>> ET.__builtins__["open"] = lambda f,m: open(f,m)
>>> ET.parse('fakeFile.xml')
jep.JepException: <type 'exceptions.IOError'>: (2, 'No such file or directory', 'fakeFile.xml')
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:647)
	at /awips2/python/lib/python2.7/xml/etree/ElementTree.parse(ElementTree.py:1182)
```

The problem with both of these solutions is that the ElementTree module may exist in multiple interpreters at the same time. Each interpreter would need to have the __builtins__ configured for that interpreter so any concurrency would cause restricted mode errors again.

Some other solutions I am considering:

1. Share __builtins__ across all jep interpreters.
1. Make a copy of each shared module for each interpreter that has a local __builtins__. This would break any modules that store internal state that needs to be shared.
1. Write a custom open() that manipulates the __builtins__ to escape restricted mode. This would probably require calling into custom c code to escape restriction. I'm not sure how to fix code unmarshalling and we would have to do the same thing anywhere else this comes up.

**Comment by bsteffensmeier** [comment]
I have a [branch](https://github.com/bsteffensmeier/jep/commit/352c461ac4d06ba6bc79bf916b657f9095892dbb) where I have proven that it is possible to share `__builtins__` in python 2.7. It must access parts of the PyInterpreterState that are not part of the public API but since it is only enabled on python2 I am not worried that the struct will change. This branch is just a proof of concept, it is probably leaking and poorly written so it needs more work before it is ready to merge in.

The main complication that I ran into is that we don't have an opportunity to replace builtins until [after the sys module is loaded](https://github.com/python/cpython/blob/2.7/Python/pythonrun.c#L609). The sys module also loads other pieces of the system so by the time we can substitute a shared builtins the existing builtins has already been spread to many places(22 places based off ob_refcnt). I was able to work around this by replacing builtins in every module loaded but there isn't a strong guarantee that there isn't another reference floating somewhere.

Another addition may be to only share builtins if the JepConfig actually uses shared modules. This would limit the damage if somehow a reference to the old builtins leaked out. It would also help preserve the idea of keeping subinterpreters independent, shared builtin really ruins the sandbox created by using sub interpreters.
